File: src/main/java/com/liamfarrell/android/snapbattle/activity/startup/ChooseNameStartupFragment.java
package com.liamfarrell.android.snapbattle.activity.startup;



import android.app.Activity;
import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.Toast;

import com.amazonaws.AmazonClientException;
import com.amazonaws.AmazonServiceException;
import com.amazonaws.mobileconnectors.lambdainvoker.LambdaFunctionException;
import com.amazonaws.mobileconnectors.lambdainvoker.LambdaInvokerFactory;
import com.amazonaws.regions.Regions;
import com.liamfarrell.android.snapbattle.model.AsyncTaskResult;
import com.liamfarrell.android.snapbattle.model.aws_lambda_function_deserialization.aws_lambda_functions.LambdaFunctionsInterface;
import com.liamfarrell.android.snapbattle.R;
import com.liamfarrell.android.snapbattle.activity.FacebookLoginFragment;
import com.liamfarrell.android.snapbattle.model.aws_lambda_function_deserialization.aws_lambda_functions.response.UpdateNameResponse;
import com.liamfarrell.android.snapbattle.model.lambda_function_request_objects.UpdateNameRequest;
import com.liamfarrell.android.snapbattle.util.HandleLambdaError;

import java.lang.ref.WeakReference;

/**
 * Created by Liam on 1/01/2018.
 */

public class ChooseNameStartupFragment extends Fragment
{
    private EditText mNameEditText;


    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_choose_name_startup, container, false);
        mNameEditText = v.findViewById(R.id.nameEditText);
        //only show next button if name entered length > 0
        mNameEditText.addTextChangedListener(new TextWatcher() {

            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                if (charSequence.length() == 0)
                {
                    ((StartupActivity)getActivity()).setEnableNextButton(true);
                }
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void afterTextChanged(Editable editable) {
                if (editable.length() == 0)
                {
                    ((StartupActivity)getActivity()).setEnableNextButton(false);
                }

            }
        });

        try
        {
            //Use the facebook name of the user as suggested name
            mNameEditText.setText(((StartupActivity)getActivity()).getName());
        }
        catch (NullPointerException e){
            ((StartupActivity)getActivity()).setEnableNextButton(false);
        }


        return v;

    }





    public String getNameEditText()
    {
        return mNameEditText.getText().toString();
    }

    public void updateName() {
        ((StartupActivity) getActivity()).showProgressSpinner();
        String newName = mNameEditText.getText().toString();
        UpdateNameRequest request = new UpdateNameRequest();
        request.setName(newName);
        new UpdateNameTask(getActivity(), this).execute(request);
    }

    private static class UpdateNameTask extends AsyncTask<UpdateNameRequest, Void, AsyncTaskResult<UpdateNameResponse>>
    {
        private WeakReference<Activity> activityReference;
        private WeakReference<ChooseNameStartupFragment> fragmentReference;

        UpdateNameTask(Activity activity, ChooseNameStartupFragment fragment)
        {
            fragmentReference = new WeakReference<>(fragment);
            activityReference = new WeakReference<>(activity);
        }

        @Override
        protected  AsyncTaskResult<UpdateNameResponse> doInBackground(UpdateNameRequest... params) {
        // Create a LambdaInvokerFactory, to be used to instantiate the Lambda proxy
        LambdaInvokerFactory factory = new LambdaInvokerFactory(
                activityReference.get().getApplicationContext(),
                Regions.US_EAST_1,
                FacebookLoginFragment.getCredentialsProvider(activityReference.get()));

        final LambdaFunctionsInterface lambdaFunctionsInterface = factory.build(LambdaFunctionsInterface.class);

                try {
                    UpdateNameResponse response = lambdaFunctionsInterface.UpdateName(params[0]);
                    return new AsyncTaskResult<>(response);
                } catch (LambdaFunctionException lfe) {
                    Log.i("ERROR", lfe.getDetails());
                    Log.i("ERROR",lfe.getStackTrace().toString());
                    lfe.printStackTrace();

                    return new AsyncTaskResult<>(lfe);
                }
                catch (AmazonServiceException ase) {
                    // invalid credentials, incorrect AWS signature, etc
                    Log.i("ERROR", ase.getErrorMessage());
                    return new AsyncTaskResult<>(ase);
                }
                catch (AmazonClientException ace) {
                    // Network issue
                    Log.i("ERROR", ace.toString());
                    return new AsyncTaskResult<>(ace);
                }
            }

            @Override
            protected void onPostExecute(AsyncTaskResult<UpdateNameResponse> asyncResult) {
                // get a reference to the activity and fragment if it is still there
                ChooseNameStartupFragment fragment = fragmentReference.get();
                Activity activity = activityReference.get();
                if (fragment == null || fragment.isRemoving()) return;
                if (activity == null || activity.isFinishing()) return;

                ((StartupActivity)activity).hideProgressSpiner();
                UpdateNameResponse result = asyncResult.getResult();
                if (asyncResult.getError() != null)
                {
                    new HandleLambdaError().handleError(asyncResult.getError(), activity,null);
                    return;

                }
                //Name Updated
                if (result != null) {
                    if (result.getResult().equals(UpdateNameResponse.getResultNameUpdated())) {
                        SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(activity.getApplicationContext());
                        sharedPref.edit().putString(FacebookLoginFragment.NAME_SHAREDPREFS, fragment.mNameEditText.getText().toString()).commit();

                        ((StartupActivity) activity).nextFragment();
                    }
                    else if(result.getResult().equals(UpdateNameResponse.getNameTooLongErrorCode()))
                    {
                        Toast.makeText(activity, R.string.name_too_long_toast, Toast.LENGTH_SHORT ).show();
                    }
                }
            }

    }





}

File: src/main/java/com/liamfarrell/android/snapbattle/mvvm_ui/startup/ChooseNameStartupFragment.kt
package com.liamfarrell.android.snapbattle.mvvm_ui.startup

import android.annotation.SuppressLint
import android.content.Context
import android.os.Bundle
import android.preference.PreferenceManager
import android.text.Editable
import android.text.TextWatcher
import android.view.*
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import androidx.navigation.fragment.findNavController
import com.liamfarrell.android.snapbattle.R
import com.liamfarrell.android.snapbattle.databinding.FragmentChooseNameStartupBinding
import com.liamfarrell.android.snapbattle.di.*
import com.liamfarrell.android.snapbattle.viewmodels.startup.ChooseNameStartupViewModel
import java.lang.ClassCastException
import javax.inject.Inject


class ChooseNameStartupFragment : Fragment(), Injectable {

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory


    private lateinit var startupActivity : SetupToolbarInterface
    private lateinit var viewModel: ChooseNameStartupViewModel
    private lateinit var defaultName : String
    private lateinit var binding: FragmentChooseNameStartupBinding

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is SetupToolbarInterface){
            startupActivity = context
            startupActivity.setTitle(resources.getStringArray(R.array.startup_activity_titles)[1])
        } else {
            throw ClassCastException() }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        setHasOptionsMenu(true)
        defaultName = arguments?.getString("defaultName") ?: ""
        binding = FragmentChooseNameStartupBinding.inflate(inflater, container, false)
        binding.lifecycleOwner = viewLifecycleOwner
        binding.nameEditText.setText(defaultName)
        binding.nameEditText.addTextChangedListener(getNameOnChangedListener())
        viewModel = ViewModelProviders.of(this, viewModelFactory).get(ChooseNameStartupViewModel::class.java)
        binding.viewModel = viewModel
        subscribeUi()
        return binding.root
    }

    private fun subscribeUi() {
        viewModel.errorMessage.observe(viewLifecycleOwner, Observer {
            it?.let{ Toast.makeText(context, it, Toast.LENGTH_SHORT).show()}
        })
    }

    @SuppressLint("ApplySharedPref")
    fun updateName() {
        val newName = binding.nameEditText.getText().toString()
        if (defaultName == newName) {
            val sharedPref = PreferenceManager.getDefaultSharedPreferences(activity!!.applicationContext)
            sharedPref.edit().putString(LoggedInFragment.NAME_SHAREDPREFS, newName).commit()
            findNavController().navigate(R.id.action_chooseNameStartupFragment_to_chooseProfilePictureStartupFragment, arguments)
        } else {
            viewModel.updateName(newName) {findNavController().navigate(R.id.action_chooseNameStartupFragment_to_chooseProfilePictureStartupFragment, arguments)}
        }
    }


    private fun getNameOnChangedListener() = object : TextWatcher{
        override fun afterTextChanged(s: Editable?) {
            s?.let {
                if (s.isNotEmpty()) {
                    startupActivity.enableNextButton()
                } else {
                    startupActivity.disableNextButton()
                }
            }
        }
        override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
        override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
    }

    override fun onOptionsItemSelected(item : MenuItem) : Boolean {
        val id = item.itemId
        if (id == R.id.action_next) {
            updateName()
        } else if (id == R.id.action_skip) {
            findNavController().navigate(R.id.action_chooseNameStartupFragment_to_chooseProfilePictureStartupFragment, arguments)
        }
        return false
    }





}